### DISCUSSION TITLE
help to fix error in IL2CPU System.Exception

### DISCUSSION THREAD

**Discussion by Toktatop** [CosmosOS/Cosmos#2739] [Q&A] [answered]
![Screenshot 2023-07-26 165721](https://github.com/CosmosOS/Cosmos/assets/122997778/bad0c2e3-031f-494a-b45d-41a6cfa3d2a9)

error in file IL2CPU System.Exception
why i have this error in that code:
using Cosmos.HAL;
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;
using Sys = Cosmos.System;

namespace python_compile
{
    public class Kernel : Sys.Kernel
    {
        public static List<string> clas = new List<string>();
        public static List<string> useclas = new List<string>();
        protected override void BeforeRun()
        {
            clas.Add("os");
            Console.Clear();
        }

        protected override void Run()
        {
            Console.Write(">> ");
            var input = Console.ReadLine();
            if (input.StartsWith("print(\"") == true && input.EndsWith("\")") == true) {
                print(input);
            }
            else if (input.StartsWith("import "))
            {
                import(input);
            }
            else
            {
                var inputcodelist = input.Split(".");
                if (clas.Contains(inputcodelist[0]))
                {
                    runothercommand(input);
                }
            }
        }
        public static void print(string input) {
            if (input.StartsWith("print(\"") == true && input.EndsWith("\")") == true)
            {
                input = input.Substring(7);
                input = input.Replace("\")", "");
                Console.WriteLine(input);
            }

        }
        public static void runothercommand(string input) {
            string[] parts = input.Split('.');
            if (parts.Length == 2 && clas.Contains(parts[0]))
            {
                string className = parts[0];
                string moduleName = parts[1];

                // Load the assembly containing the class (this assumes the class is in the same assembly as the program)
                Assembly assembly = Assembly.GetExecutingAssembly();

                // Get the type of the specified class
                Type classType = assembly.GetType(className);

                if (classType != null)
                {
                    // Create an instance of the class using Activator
                    object instance = Activator.CreateInstance(classType);

                    // Find the method within the class by its name
                    MethodInfo method = classType.GetMethod(moduleName, BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);

                    if (method != null)
                    {
                        // Execute the method on the instance if it is an instance method, or null if it is a static method
                        method.Invoke(instance, null);
                    }
                    else
                    {
                        Console.ForegroundColor = ConsoleColor.Red;
                        Console.WriteLine("Module not found in the specified class.");
                        Console.ForegroundColor = ConsoleColor.White;
                    }
                }
                else
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine("Class not found.");
                    Console.ForegroundColor = ConsoleColor.White;
                }
            }
            else
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Invalid input format. Please use the format 'class.module'. or module" + parts[0] + "not founded.");
                Console.ForegroundColor = ConsoleColor.White;
            }
        }
        public static void import(string input)
        {
            input = input.Replace("import ", "");
            string[] parts = input.Split('.');

            if (parts.Length != 2)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Error: Invalid input format. Please use the format 'class.module'.");
                Console.ForegroundColor = ConsoleColor.White;
                return;
            }

            // Extract the class name from the input
            string className = parts[0];

            // Check if the class is already imported
            if (!useclas.Contains(className))
            {
                // Load the assembly containing the class using the Cosmos assembly loader
                var asm = AppDomain.CurrentDomain.GetAssemblies();
                Assembly assembly = null;
                foreach (var a in asm)
                {
                    if (a.FullName.Contains(className))
                    {
                        assembly = a;
                        break;
                    }
                }

                if (assembly != null)
                {
                    // Add the class name to the useclas list
                    useclas.Add(className);
                }
                else
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine("Error: Class not found.");
                    Console.ForegroundColor = ConsoleColor.White;
                }
            }
        }

        public class os
        {
            public static void system(string input)
            {
                if (input == "cls") {
                    Console.Clear();
                }
            }
        }
        }
}

i use User Kit 20221121 and visual studio Comunity 2022

**Comment by retrievaaaa** [comment, marked as answer]
you are using reflection a lot; only a very small portion of Reflection is supported.

**Comment by zarlo** [comment]
we need your build logs but @GoldenretriverYT is right

**Comment by Guillermo-Santos** [comment]
Is there a place where one can see what portions of reflection are supported? Or should it be tested by seeing if it compiles?

**Reply by retrievaaaa** [reply]
Not really except for looking at the source code; but I can tell you right away that what you are trying to achieve is definitely not possible in Cosmos right now.

**Reply by retrievaaaa** [reply]
nvm your not the same person lol

but its really basic, its like base type, type name, isvaluetype, isstruct and i think thats about it
